### DISCUSSION TITLE
Errors in upgrading nuxt to v3.5.1

### DISCUSSION THREAD

**Discussion by LorenzoRottigni** [nuxt/nuxt#21080] [General]
## Initial state
This morning i was working on v3.3.2 of nuxt and i decided to upgrade to v3.5.1 in order to integrate newest features.
The application of v3.3.2 was working without any error or warning.

## Upgrade time
I upgraded nuxt using command `npx nuxi upgrade --latest` and it required me to manually update some type of some library,  for example i had to move config for @nuxt/i18n on external file i18n.confing.ts.
I followed the warnings and errors thrown and i achieved a working build.

## Testing time
Once i successfully upgraded dependencies and their confs i started to test my application and immediately i received several blocking errors.
The first kind of errors is about getSSRProps for some directive i have (ex. @vueuse v-scroll directive), by trying to remove directives from highlighted components i get errors fixed.
The second kind of errors is about composable imports on both .ts and .tsx files, for example the following imports:
`import { useLocalISO639_1Sanitizer, useLocaleSanitizer } from '~/composables/locale'`
Throws this error:
`[nitro] [dev] [unhandledRejection] TypeError: $setup.useLocalISO639_1Sanitizer is not a function                                                                                 5:25:19 PM
    at ssrRender (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/src/app.vue:27:18)
    at renderComponentSubTree (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:683:9)
    at renderComponentVNode (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:631:12)
    at Module.ssrRenderComponent (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:86:10)
    at default (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/nuxt/dist/app/components/nuxt-root.vue:76:37)
    at Module.ssrRenderSuspense (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:461:5)
    at _sfc_ssrRender (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/nuxt/dist/app/components/nuxt-root.vue:67:25)
    at renderComponentSubTree (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:683:9)
    at renderComponentVNode (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:631:12)
    at renderToString (/Users/lorenzorottigni/Documents/portfolio.rottigni.tech.v2/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:902:24)`

## Help me!
These are the errors that are actually blocking me, i don't exclude that by fixing these errors something new of strange appears.
I debugged a lot on this and i've tried with both nuxt v3.4 and v3.5, any idea of what's happening and how to fix?

**Comment by LorenzoRottigni** [comment]
Once i removed directives and composables that were throwing errors i faced a completely new error...
I have a special handlure in my project for autoimporting sections of pages based on the current route, something like this:
```
import * as components from '#components'

export const useRadar = defineStore('radar', (): RadarStore => {
const routeComponents: WritableComputedRef<RouteComponent[]> = computed(
    () => Object.entries(components)
      .filter(
        (component: [string, ConcreteComponent]): boolean => {
          const key = component?.[0]?.toLowerCase()
          return routeRaw.value?.name
            ? (
                key?.includes(
                  routeRaw.value.name.toString()
                    ?.toLowerCase()
                    ?.split('___')[0]
                    ?.replace('index', 'home'),
                )
                && !key.includes('lazy')
                && key.includes('page')
              )
            : false
        },
      )
      .map(
        (
          component: [string, ConcreteComponent],
        ): RouteComponent => ({
          component: component[1],
          letter: greekAlphabet.find(
            (letter: GreekLetter) => {
              return component[0].toLowerCase().includes(letter?.word?.toLowerCase())
            },
          ),
        }),
      )
      .sort(
        (a: RouteComponent, b: RouteComponent) => (a?.letter?.index && b?.letter?.index)
          ? (a?.letter?.index - b?.letter?.index)
          : 1,
      ),
  )

return { routeComponents }
})
```

Since version 3.5.1 i am no longer able to import components like this, i'm receiving error "cannot read properties of undefined reading 'components'". Did something change in this handlure by Nuxt or Vue side? i still have the file components.d.ts that contains the references for all components. This is just one of the things that are no longer working and i would like to understand if it's my code organization fault or something else... Looking for some help...

**Reply by danielroe** [reply]
Yes, it's no longer supported to import all components this way: see https://github.com/nuxt/nuxt/issues/20733#issuecomment-1539151092.

**Reply by LorenzoRottigni** [reply]
> Yes, it's no longer supported to import all components this way: see https://github.com/nuxt/nuxt/issues/20733#issuecomment-1539151092.

I understand, thanks for your answer... i've just structured my whole project on this kind of imports 😭😭, I have several dynamic slots that are filled based on how I name components files. I can work on this thanks! Regarding the vScroll directive from vueuse that is throwing SSRProps error, do have any idea of why? I also have troubles by importing composables.
Do you think that other errors might be a consequence of the #components import deprecation?

**Reply by danielroe** [reply]
Have you tried registering them globally as suggested in the linked issue?

**Reply by LorenzoRottigni** [reply]
> globally

I definitely solved the problem by changing part of my handlure in order to don't import components from #components. I think that this also boosted the performances and the chunks generation of my project! I can confirm that all other problems were related to #components import if someone else is facing something of similar.

Thanks @danielroe for your help 🙂.
